src/librustc_trans/mir/lvalue.rs

```diff
@@ -280,32 +273,30 @@ impl<'a, 'tcx> MirContext<'a, 'tcx> {
                 }
             }
             mir::Lvalue::Projection(ref projection) => {
-                let tr_base = self.trans_lvalue(bcx, &projection.base);
+                let mut tr_base = self.trans_lvalue(bcx, &projection.base);
                 let projected_ty = tr_base.ty.projection_ty(tcx, &projection.elem);
                 let projected_ty = self.monomorphize(&projected_ty);
 
+                let llval = tr_base.llval;
+                let ty = tr_base.ty;
                 let project_index = |llindex| {
-                    let element = if let ty::TySlice(_) = tr_base.ty.to_ty(tcx).sty {
+                    let element = if let ty::TySlice(_) = ty.to_ty(tcx).sty {
                         // Slices already point to the array element type.
-                        bcx.inbounds_gep(tr_base.llval, &[llindex])
+                        bcx.inbounds_gep(llval, &[llindex])
                     } else {
                         let zero = common::C_uint(bcx.ccx, 0u64);
-                        bcx.inbounds_gep(tr_base.llval, &[zero, llindex])
+                        bcx.inbounds_gep(llval, &[zero, llindex])
                     };
                     element
                 };
 
                 let (llprojected, llextra) = match projection.elem {
                     mir::ProjectionElem::Deref => bug!(),
                     mir::ProjectionElem::Field(ref field, _) => {
-                        let is_sized = self.ccx.shared().type_is_sized(projected_ty.to_ty(tcx));
-                        let base = if is_sized {
-                            LvalueRef::new_sized(tr_base.llval, tr_base.ty)
-                        } else {
-                            LvalueRef::new_unsized(tr_base.llval, tr_base.llextra, tr_base.ty)
-                        };
-                        let llprojected = base.trans_field_ptr(bcx, field.index());
-                        (llprojected, base.llextra)
+                        if self.ccx.shared().type_is_sized(projected_ty.to_ty(tcx)) {
+                            tr_base.llextra = ptr::null_mut();
```

Comment:
I wouldn't mutate `tr_base` here but rather have a different variable (since it ends up in `llextra` anyway).